src/imitation/algorithms/preference_comparisons.py

```diff
@@ -641,7 +642,11 @@ def __call__(
 
         # we need two fragments for each comparison
         for _ in range(2 * num_pairs):
-            traj = self.rng.choice(trajectories, p=np.array(weights) / sum(weights))
+            # NumPy's annotation here is overly-conservative, but this works at runtime
```

Comment:
same comment as above

Reply:
It looks like `numpy` is fine with lists actually, the issue is just that it expects the arrays to be of scalars:

```
(venv) adam@puffin:~/dev/imitation$ cat test.py
from numpy import random

rng : random.Generator = random.default_rng()

my_list = [1, 2, 3, 4, 5]
random.shuffle(my_list)
rng.shuffle(my_list)

class Foo:
    def __init__(self, x) -> None:
        self.x = x

my_complex_list = [Foo(42), Foo(21)]
random.shuffle(my_complex_list)
rng.shuffle(my_complex_list)
(venv) adam@puffin:~/dev/imitation$ mypy test.py
test.py:14: error: Argument 1 has incompatible type "List[Foo]"; expected "Union[_SupportsArray[dtype[Any]], _NestedSequence[_SupportsArray[dtype[Any]]], bool, int, float, complex, str, bytes, _NestedSequence[Union[bool, int, float, complex, str, bytes]]]"
test.py:15: error: Argument 1 to "shuffle" of "Generator" has incompatible type "List[Foo]"; expected "Union[_SupportsArray[dtype[Any]], _NestedSequence[_SupportsArray[dtype[Any]]], bool, int, float, complex, str, bytes, _NestedSequence[Union[bool, int, float, complex, str, bytes]]]"
Found 2 errors in 1 file (checked 1 source file)
```

I'm not sure whether they'd really want to change it given this, the one for `shuffle` is already maximally permissive within their type hierarchy taking `ArrayLike`: https://github.com/numpy/numpy/blob/266aad7478bc7fbcc55eea7f942a0d373b838396/numpy/random/_generator.pyi But `ArrayLike` is just quite restrictive: https://github.com/numpy/numpy/blob/main/numpy/_typing/_array_like.py#L85  Seems like a deliberate choice, changing that might break a lot of other annotations...

Reply:
Yeah I see, interesting, let's just deal with this internally then.